pkg/faq/faq.go

```diff
@@ -237,49 +221,53 @@ func detectFormat(file File) (formats.Encoding, error) {
 		}
 	}
 
-	fileBytes, err := file.Contents()
-	if err != nil {
-		return nil, err
-	}
+	reader := file.Reader()
+	// Look for either {, <, or --- at the beginning of the file to detect
+	// json/xml/yaml.
 
-	format := strings.ToLower(linguist.Analyse(fileBytes, linguist.LanguageHints(file.Path())))
-
-	// If linguist doesn't detect care about then try to take a better guess.
-	if _, ok := formats.ByName(format); !ok {
-		// Look for either {, <, or --- at the beginning of the file to detect
-		// json/xml/yaml.
-		scanner := bufio.NewScanner(bytes.NewReader(fileBytes))
-		keepScanning := true
-		for keepScanning && scanner.Scan() {
-			line := scanner.Bytes()
-			for i, b := range line {
-				// Go through each byte until we find a non-whitespace
-				// character.
-				if !unicode.IsSpace(rune(b)) {
-					// If it's either character we're looking for, set the
-					// correct format.
-					if b == '{' {
-						format = "json"
-					} else if b == '<' {
-						format = "xml"
-					} else if b == '-' {
-						// If we run into a -, then check if there is a yaml
-						// document separator ---.
-						if len(line[i:]) >= 3 && bytes.Equal(line[i:i+3], yamlSeparator) {
-							format = "yaml"
-						}
-					}
-					// Break here because if the first non-whitespace character
-					// isn't what we're looking for, then we didn't find what
-					// we're looking for.
-					keepScanning = false
-					break
-				}
+	var format string
+	for peekN := 1; ; peekN++ {
+		fileBytes, err := reader.Peek(peekN)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		b := fileBytes[peekN-1]
+
+		// whitespace can be ignored
+		if unicode.IsSpace(rune(b)) {
+			continue
+		}
+
+		// If it's any of the characters we're looking for, set the
+		// correct format.
+		if b == '{' {
+			format = "json"
+			break
+		} else if b == '<' {
+			format = "xml"
+			break
+		} else if b == '-' {
+			// If we run into a -, then check if there is a yaml
+			// document separator ---.
+			fileBytes, err = reader.Peek(peekN + 2)
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+			potentialYamlBytes := fileBytes[peekN-1 : peekN+2]
+			if bytes.Equal(potentialYamlBytes, yamlSeparator) {
+				format = "yaml"
+				break
 			}
 		}
+		// we found a non-whitespace character that isn't what we were
+		// looking for, so stop trying to detect the format.
```

Comment:
Full sentences with caps and punctuation!

Reply:
Sure would be nice if the linter handled this!